tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,310 @@
+    }
+
+    fn set_parent(&self, current: &Id, parent: &Id) {
+        // TODO(david): remove `.expect()`; return an error.
+        let current_span = self
+            .spans
+            .get(id_to_idx(current))
+            .expect(&format!("Missing current span with id: {:?}", current));
+        let parent_span = self
+            .spans
+            .get(id_to_idx(parent))
+            .expect(&format!("Missing parent span with id: {:?}", parent));
+
+        // current_span.parent = Some(parent.clone());
+        // parent_span.children.push(current.clone())
+    }
+
+    /// Decrements the reference count of the span with the given `id`, and
+    /// removes the span if it is zero.
+    ///
+    /// The allocated span slot will be reused when a new span is created.
+    fn drop_span(&self, id: Id) -> bool {
+        let idx = id_to_idx(&id);
+        if !self
+            .spans
+            .get(idx)
+            .and_then(|span| Some(span.drop_ref(Ordering::Relaxed)))
+            .unwrap_or_else(|| {
+                panic!("tried to drop {:?} but it no longer exists!", id);
+            })
+        {
+            return false;
+        }
+
+        // Synchronize only if we are actually removing the span (stolen
+        // from std::Arc);
+        fence(Ordering::Acquire);
+        self.spans.remove(idx)
+    }
+}
+
+thread_local! {
+    static CURRENT_SPANS: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+impl Subscriber for Registry {
+    fn register_callsite(&self, _: &'static Metadata<'static>) -> Interest {
+        Interest::always()
+    }
+
+    fn enabled(&self, _: &Metadata<'_>) -> bool {
+        true
+    }
+
+    #[inline]
+    fn new_span(&self, attrs: &span::Attributes<'_>) -> span::Id {
+        let mut values = HashMap::new();
+        let mut visitor = RegistryVisitor(&mut values);
+        attrs.record(&mut visitor);
+        let parent = attrs.parent().map(|id| id.clone());
+        let s = Data {
+            metadata: attrs.metadata(),
+            parent,
+            children: vec![],
+            ref_count: AtomicUsize::new(1),
+            values: Mutex::new(values),
+        };
+        let id = self.insert(s).expect("Unable to allocate another span");
+        idx_to_id(id)
+    }
+
+    #[inline]
+    fn record(&self, span: &span::Id, values: &span::Record<'_>) {
+        dbg!(span);
+        dbg!(values);
+    }
+
+    fn record_follows_from(&self, span: &span::Id, follows: &span::Id) {
+        let current = self
+            .get(span)
+            .expect("Current span is missing; this is a bug");
+
+        let current = self
+            .span(span)
+            .expect("Current span is missing; this is a bug");
+    }
+
+    fn enter(&self, id: &span::Id) {
+        CURRENT_SPANS.with(|spans| {
+            spans.borrow_mut().push(id.clone());
+            if let Some(data) = self.get(id) {
+                data.ref_count.fetch_add(1, Ordering::SeqCst);
```

Comment:
Why does this need to be sequentially consistent? The adds can all be `Relaxed`.

Also, why not just replace this whole `if let` with a call to `self.clone_span(id)`?

Reply:
Fixed on both! Thanks.